presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ReplaceTopNWithSort.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.SortNode;
+import io.prestosql.sql.planner.plan.TopNNode;
+
+import static io.prestosql.sql.planner.optimizations.QueryCardinalityUtil.isScalar;
+import static io.prestosql.sql.planner.plan.Patterns.TopN.count;
+import static io.prestosql.sql.planner.plan.Patterns.topN;
+
+/**
+ *  Replace TopN node with a Sort node when the subplan is guaranteed to produce fewer rows than the TopN limit
+ */
+public class ReplaceTopNWithSort
+        implements Rule<TopNNode>
+{
+    private static final Pattern<TopNNode> PATTERN = topN().with(count().matching(count -> count != 0));
+
+    @Override
+    public Pattern<TopNNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TopNNode node, Captures captures, Context context)
+    {
+        if (isScalar(node.getSource(), context.getLookup())) {
+            return Result.ofPlanNode(new SortNode(context.getIdAllocator().getNextId(), node.getSource(), node.getOrderingScheme()));
```

Comment:
No need to sort single node. I would extract plan cardinality and if it is 1 then remove TopN is lower than `TopN::count` then I would replace it with `Sort`

Reply:
But ideally it would be handled by the `RemoveSingleRowSort` rule right ? I assumed it was redundant to add it here.

Reply:
Yes, but inlining this here would save us one iterative optimizer loop. Also I think you could handle the case with limit = 0 here.

However, as you pointed it could be a matter of taste. Up to you.